fuel-bft/src/round.rs

```diff
@@ -38,39 +63,108 @@ impl fmt::Display for HeightRound {
 }
 
 impl HeightRound {
+    /// Genesis round
+    pub const INITIAL: Self = Self::new(0, 0);
+
+    /// Infinity round
+    pub const NEVER: Self = Self::new(u64::MAX, u64::MAX);
+
+    /// Create a new tuple height+round
     pub const fn new(height: u64, round: u64) -> Self {
         Self { height, round }
     }
 
+    /// Start a new round in the provided height
     pub const fn start(height: u64) -> Self {
-        Self::new(height, 0)
+        Self::new(height, Self::INITIAL.round)
     }
 
+    /// Block height
     pub const fn height(&self) -> u64 {
         self.height
     }
 
+    /// Height round step.
     pub const fn round(&self) -> u64 {
         self.round
     }
 
+    /// Decrement the height, returning [`Self::NEVER`] in case of underflow
+    pub const fn previous_height(&self) -> u64 {
+        self.height.wrapping_sub(1)
+    }
+
+    /// Decrement the round, returning [`Self::NEVER`] in case of underflow
+    pub const fn previous_round(&self) -> u64 {
+        self.round.wrapping_sub(1)
+    }
+
+    /// Increment the height, resetting the round
     pub const fn increment_height(self) -> Self {
         Self {
             height: self.height + 1,
-            round: 0,
+            round: Self::INITIAL.round,
         }
     }
 
+    /// Increment the round without mutation the height
     pub const fn increment_round(self) -> Self {
         Self {
             height: self.height,
             round: self.round + 1,
         }
     }
+
+    /// Check if the round is in genesis.
+    pub const fn is_initial(&self) -> bool {
+        self.height == Self::INITIAL.height && self.round == Self::INITIAL.round
+    }
+
+    /// Check if the round is in infinity.
+    pub const fn is_never(&self) -> bool {
+        self.height == Self::NEVER.height && self.round == Self::NEVER.round
+    }
 }
 
 impl From<u64> for HeightRound {
     fn from(height: u64) -> Self {
         Self::start(height)
     }
 }
+
+/// Tuple containing a height, round and validator publicn key
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct RoundValidator {
+    round: HeightRound,
+    validator: PublicKey,
+}
+
+impl RoundValidator {
+    /// Create a new instance.
+    pub const fn new(round: HeightRound, validator: PublicKey) -> Self {
+        Self { round, validator }
+    }
+
+    /// Height and round.
+    pub const fn round(&self) -> &HeightRound {
+        &self.round
+    }
+
+    /// Public key of the validator.
+    pub const fn validator(&self) -> &PublicKey {
+        &self.validator
+    }
+
+    /// Replace the validator public key.
+    pub fn set_validator(&mut self, validator: PublicKey) {
+        self.validator = validator;
+    }
+}
+
+impl From<HeightRound> for RoundValidator {
```

Comment:
What is the purpose of this conversion? It seems lossy?

Reply:
We need the ability to create unsigned messages for anonymous authors. The author is assigned after the message is signed, not before.

However, I'm refactoring this bit of code in a PR that will follow this one. Currently the messages are the same for every round step, and they should be dedicated variants instead.